### DISCUSSION TITLE
Mutating webhook is failing sometimesand causing issue in creation of other pvcs

### DISCUSSION THREAD

**Discussion by richa-mishra-aera** [topolvm/pvc-autoresizer#293] [Q&A]
I have installed autoresizer and observed that there is a mutating webhook pvc-autoresizer-mutating-webhook-configuration which is failing sometimes and causing issue for creating other pvcs. Attaching screenshot.
can you please suggest here. And if we set the `.Values.webhook.pvcMutatingWebhook.enabled` as false, then what impact it will have?
![image (24)](https://github.com/user-attachments/assets/48a8efd4-6d63-471d-95b1-43b2f51759d1)

**Comment by pluser** [comment]
Hi,

Regarding the webhook error, first of all, please check whether you can create PVC and Pod normally. If it seems to be normal, it is likely that the certificate installation has failed, so it would be good to check whether cert-manager is working correctly.

Regarding your question about what happens when you disable webhooks, the initial expansion will be disabled. For more information, please see [this proposal](https://github.com/topolvm/pvc-autoresizer/blob/f794d6fe38804d77954a6b02fa21d472a5b34401/docs/proposals/resize-when-creating-by-group.md). If you are not using this feature, you can disable it.

**Reply by richa-mishra-aera** [reply]
The issue is intermittent, if i try to create a normal pvc, i get this error

**Reply by pluser** [reply]
hmm... I am suspicious that the settings by cert-manager are not working properly.

Please check that `metadata.annotations.cert-manager.io/inject-ca-from` and `webhooks[0].clientConfig.caBundle` are present in the MutatingWebhookConfiguration resource.

```
$ kubectl get MutatingWebhookConfiguration -n pvc-autoresizer mutating-webhook-configuration -o yaml
apiVersion: admissionregistration.k8s.io/v1
kind: MutatingWebhookConfiguration
metadata:
  annotations:
    cert-manager.io/inject-ca-from: pvc-autoresizer/pvc-autoresizer-controller # should be exists
  name: mutating-webhook-configuration
webhooks:
- admissionReviewVersions:
  - v1
  clientConfig:
    caBundle: LS0t... # should be exists
...
```

Also, please check that the secret for the webhook has actually been copied correctly.

```
$ kubectl -n pvc-autoresizer get secret pvc-autoresizer-webhook-ca -ojsonpath='{.data.ca\.crt}'   | base64 -d | openssl x509 -noout -text
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            ee:8f:4f:c8:55:7b:16:76:d8:6a:a2:e5:94:bc:7c:6b
        Issuer: CN = ca.webhook.pvc-autoresizer
        Subject: CN = ca.webhook.pvc-autoresizer
...
```

should be equals to

```
$ kubectl get mutatingwebhookconfigurations mutating-webhook-configuration -ojson   | jq '.webhooks[].clientConfig.caBundle' -r | base64 -d   | openssl x509 -noout -text
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            ee:8f:4f:c8:55:7b:16:76:d8:6a:a2:e5:94:bc:7c:6b
        Issuer: CN = ca.webhook.pvc-autoresizer
        Subject: CN = ca.webhook.pvc-autoresizer
...
```

**Reply by akash-warkhade-aera** [reply]
@pluser We have checked above things, caBundle is present in MutatingWebhookConfiguration named **pvc-autoresizer-mutating-webhook-configuration**.
certificate as well as issuer in pvc-autoresizer-webhook-ca secret is different than caBundle of Mutatingwebhookconfiguration
for webhook-ca below is the cert :
![image](https://github.com/user-attachments/assets/ade3cf9f-80d8-4c6c-8184-e44ac59b08e4)

CaBundle in mutatingWebhookConfiguration has below cert:
![image](https://github.com/user-attachments/assets/403030c7-565b-46cf-9868-db3b76933081)

**Reply by akash-warkhade-aera** [reply]
@pluser  can you please help on above

**Reply by akash-warkhade-aera** [reply]
@pluser Can you please check provided details and help identifying  the issue?

**Reply by akash-warkhade-aera** [reply]
hi @pluser 
do you have any input on above ? please let us know how we can fix it ?

**Reply by pluser** [reply]
It's strange that the contents of caBundle and secret don't match. It seems that cert-manager is not working properly. Could you check that injector is working properly and check its log?
```
$ kubectl -n cert-manager get pods -l app.kubernetes.io/name=cainjector
...
$ kubectl logs -n cert-manager -l app.kubernetes.io/name=cainjector
```

**Reply by akash-warkhade-aera** [reply]
@pluser  cainjector pod & logs looking fine

$ kubectl -n cert-manager get pods -l app.kubernetes.io/name=cainjector
NAME                                       READY   STATUS    RESTARTS   AGE
cert-manager-cainjector-79674d85c9-wdfrw   1/1     Running   0          15d


$ kubectl logs -n cert-manager -l app.kubernetes.io/name=cainjector

I0120 08:33:27.620631       1 reflector.go:800] k8s.io/client-go@v0.29.0/tools/cache/reflector.go:229: Watch close - *v1.Secret total 255 items received
I0120 08:33:27.623096       1 round_trippers.go:553] GET https://172.16.0.1:443/api/v1/secrets?allowWatchBookmarks=true&resourceVersion=107615282&timeoutSeconds=444&watch=true 200 OK in 2 milliseconds
I0120 08:33:27.896922       1 round_trippers.go:553] GET https://172.16.0.1:443/apis/coordination.k8s.io/v1/namespaces/cert-manager/leases/cert-manager-cainjector-leader-election 200 OK in 4 milliseconds
I0120 08:33:27.902044       1 round_trippers.go:553] PUT https://172.16.0.1:443/apis/coordination.k8s.io/v1/namespaces/cert-manager/leases/cert-manager-cainjector-leader-election 200 OK in 4 milliseconds
I0120 08:33:27.902148       1 leaderelection.go:281] successfully renewed lease cert-manager/cert-manager-cainjector-leader-election
I0120 08:33:42.906906       1 round_trippers.go:553] GET https://172.16.0.1:443/apis/coordination.k8s.io/v1/namespaces/cert-manager/leases/cert-manager-cainjector-leader-election 200 OK in 4 milliseconds
I0120 08:33:42.912038       1 round_trippers.go:553] PUT https://172.16.0.1:443/apis/coordination.k8s.io/v1/namespaces/cert-manager/leases/cert-manager-cainjector-leader-election 200 OK in 4 milliseconds
I0120 08:33:42.912145       1 leaderelection.go:281] successfully renewed lease cert-manager/cert-manager-cainjector-leader-election
I0120 08:33:45.689245       1 reflector.go:800] k8s.io/client-go@v0.29.0/tools/cache/reflector.go:229: Watch close - *v1.MutatingWebhookConfiguration total 99 items received
I0120 08:33:45.690578       1 round_trippers.go:553] GET https://172.16.0.1:443/apis/admissionregistration.k8s.io/v1/mutatingwebhookconfigurations?allowWatchBookmarks=true&resourceVersion=107614557&timeoutSeconds=360&watch=true 200 OK in 1 milliseconds

**Reply by pluser** [reply]
Do you have a `cert-manager.io/inject-ca-from: pvc-autoresizer/pvc-autoresizer-controller` annotation in MutatingWebhookConfiguration?

```
kubectl get MutatingWebhookConfiguration -n pvc-autoresizer pvc-autoresizer-mutating-webhook-configuration -o yaml | grep cert-manager.io/inject-ca-from
```

**Reply by akash-warkhade-aera** [reply]
@pluser It doesn't look like it has.
I even tried searching for cert.its not there also 
`$ kubectl get MutatingWebhookConfiguration -n pvc-autoresizer pvc-autoresizer-mutating-webhook-configuration -o yaml|grep -i cert
`

**Reply by pluser** [reply]
If that annotation doesn't exist, that's probably part of the cause. It should exist.

What would happen if you set that annotation manually?

If you're installing from helm chart, please check the values. In particular, please check that `webhook.caBundle` and `webhook.certificate.generate`.

**Reply by akash-warkhade-aera** [reply]
@pluser  We are using default values for both of these parameters. 
for webhook.certificate.generate its set to true
what should be the value for webhook.caBundle ?

**Reply by pluser** [reply]
`webhook.certificate.generate` [should be false by default](https://github.com/topolvm/pvc-autoresizer/blob/0f33e386b10b22207998933a96f609bf46f27b73/charts/pvc-autoresizer/README.md?plain=1#L68). And caBundle is nil by default.
Please try setting the `cert-manager.io/inject-ca-from` annotation manually.

**Reply by akash-warkhade-aera** [reply]
ok where we need to set this in helm chart ? and what should be the value of cert-manager.io/inject-ca-from annotation ? @pluser

**Reply by akash-warkhade-aera** [reply]
@pluser does it need to set to certificate from cert-manager ?  I don't see any certificate in cert-manager namespace
$ kubectl get certificates -n cert-manager
No resources found in cert-manager namespace.

**Reply by pluser** [reply]
Certificate should exist in the pvc-autoresizer namespace.

What value should be set in the annotation to MutatingWebhookConfiguration depends on your environment. Maybe `cert-manager.io/inject-ca-from=pvc-autoresizer/pvc-autoresizer-controller`?

Please also check the cert-manager documentation: https://cert-manager.io/docs/concepts/ca-injector/#injecting-ca-data-from-a-certificate-resource

**Reply by akash-warkhade-aera** [reply]
@pluser  we have updated webhook.certificate.generate to false and commented out caBundle
now its automatically added cert-manager.io/inject-ca-from": "pvc-autoresizer/pvc-autoresizer-controller
so should we keep it like this or manually add cert-manager.io/inject-ca-from=pvc-autoresizer/pvc-autoresizer-controller in helm chart ?
if so in helm values where should we add it

**Reply by akash-warkhade-aera** [reply]
@pluser  please update

**Reply by pluser** [reply]
If the annotation was inserted automatically, then it's OK.
Does the `x509: certificate signed by unknown authority` error still occur?

**Reply by akash-warkhade-aera** [reply]
no now it doesn't occur

**Reply by pluser** [reply]
That's good. Can I close this thread then?

**Reply by toshipp** [reply]
This discussion has been closed due to inactivity. Please feel free to reopen this discussion (or open a new one) if you still need to discuss this topic. Thank you for your contribution.
